File: Data Structures & Algorithms/binary-search/submission-0.java
class Solution {
    public int search(int[] nums, int target) {
        int mid = nums.length / 2 - 1;
        return recursiveBin(0, nums.length - 1, nums, target );

    }

    public int recursiveBin(int l, int r, int[] nums, int target) {
        if (l > r) return -1;
        int mid = l + (r - l) / 2;
        if (nums[mid] == target) return mid;
        else if (nums[mid] > target) {
            int newMid = (nums.length - 1 + mid) / 2;
            return recursiveBin(l, mid - 1, nums, target);
        }
        else {
            int newMid = mid / 2;
            return recursiveBin(mid + 1, r, nums, target);
        }
    }
}

File: Data Structures & Algorithms/find-minimum-in-rotated-sorted-array/submission-0.py
class Solution:
    def findMin(self, nums: List[int]) -> int:
        # base case
        if len(nums) == 1: return nums[-1]

        # binary search, pointers at ends
        l = 0
        r = len(nums) - 1
        while (l < r):
            # check if we are at the lowest two numbers
            if (r - l < 2):
                if (nums[r] < nums[l]):
                    return nums[r]
                else:
                    return nums[l]

            # determine if mid -> r is sorted
            mid = l + int((r - l) / 2)
            if (nums[r] > nums[mid]):
                # if mid -> r is strictly increasing, the min cannot be in there, so r = mid
                r = mid
            if (nums[r] < nums[mid]):
                # if mid -> r is not strictly increasing, the min is in there so we look between there
                l = mid + 1

        return nums[l]